rust/agama-software/src/service.rs

```diff
@@ -177,6 +179,92 @@ impl Service {
 
         Ok(())
     }
+
+    /// Updates the proposal and the service state.
+    ///
+    /// This function performs the following actions:
+    ///
+    /// 1. Calculates the [wanted state](SoftwareState) using the current product, configuration,
+    ///    system information and product selection.
+    /// 2. Synchronizes the packaging system (through the model adapter).
+    /// 3. Emits issues if something is wrong.
+    /// 4. Updates the service state (system information and proposal).
+    ///
+    /// Options from 2 to 4 might take some time, so they run in a separate Tokio task.
+    async fn update_proposal(&mut self) -> Result<(), Error> {
+        let Some(product) = &self.product else {
+            return Ok(());
+        };
+
+        let product = product.read().await.clone();
+
+        let new_state = {
+            let state = self.state.read().await;
+            SoftwareState::build_from(&product, &state.config, &state.system, &self.selection)
+        };
+
+        tracing::info!("Wanted software state: {new_state:?}");
+        let model = self.model.clone();
+        let progress = self.progress.clone();
+        let issues = self.issues.clone();
+        let state = self.state.clone();
+        let events = self.events.clone();
+
+        tokio::task::spawn(async move {
+            let mut my_model = model.lock().await;
+            my_model.set_product(product);
```

Comment:
do we need to always set product? I am not sure how time consuming the operation is.

Reply:
It takes no time. Initially, I thought about removing this function, but one thing at a time :-)